Design-Docs/standardizing-cluster-infrastructure.md

```diff
@@ -0,0 +1,277 @@
+
+##### Layer 1: Pre-cluster state
+
+The pre-cluster state consists of the node image that has to be compatible with the provider and impacts both the second and third layers. This is why it has to be created for each stack and infrastructure provider. The SCS project can build frameworks that allow providers to efficiently create and publish the node images.
+
+It is important to consider that providers use different ways of providing these images. For example, Hetzner Cloud provides them indirectly by using snapshots in the creation process of a server. OpenStack, on the other hand, uses qcow2 images that are provided in a direct way. Note that in both cases images can be created with certain tools (e.g. Packer).
+
+The node images can be tested with SCS E2E tests and a test framework that can be implemented by every provider.
+
+There are various components, tools, and possible configurations that play a role in the pre-cluster state. This is a list that should serve as a reference:
+
+- Node problem detector (https://github.com/kubernetes/node-problem-detector): It depends on the OS because node errors do.
+- konnectivity-server ([https://kubernetes.io/docs/tasks/extend-kubernetes/setup-konnectivity](https://kubernetes.io/docs/tasks/extend-kubernetes/setup-konnectivity/)): kube-api-server proxy for traffic from control planes to cluster
+- Production-readiness: log-collector, log-rotator, pinning versions, DNS, NTP, ssh-hardening
+- Container Runtime (CRI-conform): cri-o, containerd
+- Low-Level Container Runtime (OCI-conform): crun, runc, runsc (gvisor), kata containers. With the help of runtime objects, this could be chosen directly in Kubernetes. Kata containers, for example, don’t work with virtualized systems without nested KVM support.
+- cgroups V2 vs V1, croupfs vs systemd
+- eBPF support: requires kernel >5.10
+- iptables requirements: depends on whether kube-proxy is used, which is not the case in some modern systems
+- SSH hardening
+- Firewall: local software firewall, externally or via cilium eBPF firewall
+
+##### Layer 2: Cluster state
+
+Cluster state consists of standardized Cluster API objects, such as `ClusterClass`, and other life cycle objects, e.g. `MachineDeployment` and `MachineHealthCheck`. Additionally, there are provider-specific objects that exist in every provider integration, like `ProviderCluster` and `ProviderMachineTemplate`, as well as other objects.
+
+SCS has to define the inputs for those objects in each Cluster Stack in such a way, that each provider can implement it.
+
+The most important standardization, but also the one that restricts the options to define clusters the most, is related to how the Kubernetes system is supposed to look like. Note that the Kubernetes system is heavily influenced by the components chosen for the pre-cluster state.
+
+The first decision that has to be taken is to select a bootstrap provider: Kubeadm or Talos. Talos would require a node image with Talos OS.
```

Comment:
What about completely custom bootstrapping? What is the usecase for offering a switch between kubeadm and talos?

Reply:
The same here. Why limit ourself to certain tools?

Reply:
We just thought about those two because they are the two most popular ones. If you know others, then please add them!